ledger/src/advance.rs

```diff
@@ -147,28 +147,24 @@ where
         }
 
         // Split off a chunk of the candidate solutions.
-        let candidates_chunk = if candidate_solutions.len() > chunk_size {
+        let mut candidates_chunk = if candidate_solutions.len() > chunk_size {
             candidate_solutions.split_off(candidate_solutions.len() - chunk_size)
         } else {
             std::mem::take(&mut candidate_solutions)
         };
 
         // Verify the solutions in the chunk.
-        let verification_results: Vec<_> = candidates_chunk
-            .into_iter()
-            .rev()
-            .map(|solution| {
-                let verified = verification_fn(&solution);
-                (solution, verified)
-            })
-            .collect();
+        let verification_results = candidates_chunk.iter_mut().rev().map(|solution| {
+            let verified = verification_fn(solution);
+            (solution, verified)
+        });
 
         // Process the results of the verification.
         for (solution, is_valid) in verification_results.into_iter() {
             if is_valid && valid_candidate_solutions.len() < max_solutions {
-                valid_candidate_solutions.push(solution);
+                valid_candidate_solutions.push(*solution);
             } else {
-                aborted_candidate_solutions.push(solution);
+                aborted_candidate_solutions.push(*solution);
             }
         }
     }
```

Comment:
Could this not be done in the same iterator instead of iterating twice? Maybe the compiler is smart enough to simplify this but seems wasteful to me.

Reply:
It's possible, but we would likely have to refactor `verification_fn`. It also isn't in the scope of this PR, so ideally any optimizations should be done in a future PR.